src/vs/editor/contrib/gotoSymbol/goToSymbol.ts

```diff
@@ -37,7 +39,17 @@ function getLocationLinks<T>(
 				result.push(value);
 			}
 		}
-		return result;
+
+		const uniqueResults: Set<LocationLink> = new Set();
+		let last: LocationLink | undefined;
+		// Preserve result order by sorting a copy and using that to compare with Set contents.
+		for (const link of [...result].sort(compareLocations)) {
+			if (last === undefined || compareLocations(last, link) !== 0) {
+				uniqueResults.add(link);
+				last = link;
+			}
+		}
+		return result.filter(link => uniqueResults.has(link));
```

Comment:
You are very, very close the "real" `ReferencesModel`. So, I'd suggest that you create an instance which does all the sorting and de-duping and that you take its result, e.g `ReferencesModel#references`. It might need some tweak but I'd say that's better than the duplication

Reply:
Let me know if you disagree and think it would still be better to create a `ReferenceModel` instance, but here were my two reasons for not doing that:

I felt that creating a `ReferenceModel` was a bit overkill here since the only two things I needed from it was `compareLocations` and a de-duplicated result array. The duplication of the sorting and comparing code only totals 4 lines and I felt it was much clearer to have that small bit of logic duplicated here rather than to use a whole `ReferenceModel` class just for this simple task.

Secondly (and maybe more importantly), there are tests that require the results from here be returned in the order of the registered providers. That's why I created the `Set` of `uniqueResults` and then filtered the original results rather than simply returning a sorted array of unique locations like the `ReferenceModel` does:
https://github.com/microsoft/vscode/blob/c6d6bda7197d0546d21465dc7f6fa849a146504f/src/vs/workbench/test/browser/api/extHostLanguageFeatures.test.ts#L524-L544

Reply:
> Secondly (and maybe more importantly), there are tests that require the results from here be returned in the order of the registered providers.

I have no problem with that. We are changing how this command/function works, so it's logical that some tests need to be adjusted.

> I felt that creating a ReferenceModel was a bit overkill here since the only two things I needed from it was compareLocations and a de-duplicated result array.

I don't think it's a big issue. At a minimum you should make `ReferencesModel._compareReferences` accessible and than reuse that

Reply:
👍  Done. Sorting and de-duplication code is now shared between `goToSymbol` and `ReferencesModel` and the tests have been adjusted.